pkg/k8s/service_cache.go

```diff
@@ -0,0 +1,384 @@
+// ServiceCache. Returns the ServiceID unless the Kubernetes endpoints could not
+// be parsed and a bool to indicate whether the endpoints was changed in the
+// cache or not.
+func (s *ServiceCache) UpdateEndpoints(k8sEndpoints *v1.Endpoints) (ServiceID, *Endpoints) {
+	svcID, newEndpoints := ParseEndpoints(k8sEndpoints)
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if oldEndpoints, ok := s.endpoints[svcID]; ok {
+		if oldEndpoints.DeepEquals(newEndpoints) {
+			return svcID, newEndpoints
+		}
+	}
+
+	s.endpoints[svcID] = newEndpoints
+
+	// Check if the corresponding Endpoints resource is already available
+	service, ok := s.services[svcID]
+	if ok {
+		s.Events <- ServiceEvent{
+			Action:       UpdateService,
+			ID:           svcID,
+			Service:      service,
+			Endpoints:    newEndpoints,
+			K8sEndpoints: k8sEndpoints,
+		}
+	}
+
+	return svcID, newEndpoints
+}
+
+// DeleteEndpoints parses a Kubernetes endpoints and removes it from the
+// ServiceCache
+func (s *ServiceCache) DeleteEndpoints(k8sEndpoints *v1.Endpoints) ServiceID {
+	svcID := ParseEndpointsID(k8sEndpoints)
+
+	s.mutex.Lock()
+	service, serviceOK := s.services[svcID]
+	endpoints, endpointsOK := s.endpoints[svcID]
+	delete(s.endpoints, svcID)
+	s.mutex.Unlock()
+
+	if serviceOK && endpointsOK {
+		s.Events <- ServiceEvent{
+			Action:    DeleteService,
+			ID:        svcID,
+			Service:   service,
+			Endpoints: endpoints,
+		}
+	}
+
+	return svcID
+}
+
+// ListMissingEndpoints returns the list of required endpoints that is not in
+// the cache
+func (s *ServiceCache) ListMissingEndpoints(requiredEndpoints versioned.Map) versioned.Map {
+	type parsedEndpoints struct {
+		id        ServiceID
+		endpoints *Endpoints
+		uuid      versioned.UUID
+		object    versioned.Object
+	}
+
+	// parse endpoints first to avoid holding the loadBalancer mutex
+	parsed := make([]parsedEndpoints, 0, len(requiredEndpoints))
+	for uuid, endpointsObj := range requiredEndpoints {
+		id, endpoints := ParseEndpoints(endpointsObj.Data.(*v1.Endpoints))
+		parsed = append(parsed, parsedEndpoints{
+			id:        id,
+			endpoints: endpoints,
+			uuid:      uuid,
+			object:    endpointsObj,
+		})
+	}
+
+	missing := versioned.NewMap()
+
+	s.mutex.RLock()
+	for _, p := range parsed {
+		existingEndpoint, ok := s.endpoints[p.id]
+		if !ok {
+			missing.Add(p.uuid, p.object)
+			continue
+		}
+		if !p.endpoints.DeepEquals(existingEndpoint) {
+			missing.Add(p.uuid, p.object)
+		}
+	}
+	s.mutex.RUnlock()
+
+	return missing
+}
+
+// UpdateIngress parses a Kubernetes ingress and adds or updates it in the
+// ServiceCache.
+func (s *ServiceCache) UpdateIngress(ingress *v1beta1.Ingress, host net.IP) (ServiceID, error) {
+	svcID, newService, err := ParseIngress(ingress, host)
+	if err != nil {
+		return svcID, err
+	}
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if oldService, ok := s.ingresses[svcID]; ok {
+		if oldService.DeepEquals(newService) {
```

Comment:
Create either a GH issue of a function that knows how to perform the deep equal of v1beta1.Ingress for the fields that actually matter to us.